File: fe/ConnectEXE/src/services/ProjectService.ts
import { API_BASE_URL } from '../constants/ApiConst';
import type { ProjectResponse, ProjectDetailResponse, ProjectListResponse } from '../types/response/ProjectResponseDTO';
import type { CreateProjectRequest, UpdateProjectRequest } from '../types/request/ProjectRequestDTO';

// Helper to get headers with auth token
const getAuthHeaders = (): HeadersInit => {
  const token = localStorage.getItem('accessToken');
  const headers: HeadersInit = {
    'Content-Type': 'application/json',
  };
  
  if (token) {
    headers['Authorization'] = `Bearer ${token}`;
  }
  
  return headers;
};

// Get random projects from subscribed users
export const getRandomProjects = async (limit: number = 6): Promise<ProjectResponse[]> => {
  const response = await fetch(`${API_BASE_URL}/projects/random?limit=${limit}`, {
    method: 'GET',
    credentials: 'include',
  });
  if (!response.ok) throw new Error('Failed to fetch random projects');
  const result = await response.json();
  return result.data || result;
};

// Get successful projects (featured)
export const getSuccessfulProjects = async (page: number = 0, size: number = 10): Promise<ProjectListResponse> => {
  const response = await fetch(`${API_BASE_URL}/projects/successful?page=${page}&size=${size}`, {
    method: 'GET',
    credentials: 'include',
  });
  if (!response.ok) throw new Error('Failed to fetch successful projects');
  const result = await response.json();
  // Backend returns Page object in data, convert to ProjectListResponse
  const pageData = result.data || result;
  return {
    projects: pageData.content || [],
    totalPages: pageData.totalPages || 0,
    totalElements: pageData.totalElements || 0,
    currentPage: pageData.number || 0,
  };
};

// Get projects by category
export const getProjectsByCategory = async (
  category: string,
  page: number = 0,
  size: number = 10
): Promise<ProjectListResponse> => {
  const response = await fetch(`${API_BASE_URL}/projects/category/${category}?page=${page}&size=${size}`, {
    method: 'GET',
    credentials: 'include',
  });
  if (!response.ok) throw new Error('Failed to fetch projects by category');
  const result = await response.json();
  const pageData = result.data || result;
  return {
    projects: pageData.content || [],
    totalPages: pageData.totalPages || 0,
    totalElements: pageData.totalElements || 0,
    currentPage: pageData.number || 0,
  };
};

// Get all projects with pagination
export const getAllProjects = async (page: number = 0, size: number = 10): Promise<ProjectListResponse> => {
  const response = await fetch(`${API_BASE_URL}/projects?page=${page}&size=${size}`, {
    method: 'GET',
    credentials: 'include',
  });
  if (!response.ok) throw new Error('Failed to fetch projects');
  const result = await response.json();
  const pageData = result.data || result;
  return {
    projects: pageData.content || [],
    totalPages: pageData.totalPages || 0,
    totalElements: pageData.totalElements || 0,
    currentPage: pageData.number || 0,
  };
};

// Get project detail
export const getProjectDetail = async (projectId: number): Promise<ProjectDetailResponse> => {
  const response = await fetch(`${API_BASE_URL}/projects/${projectId}`, {
    method: 'GET',
    credentials: 'include',
  });
  if (!response.ok) throw new Error('Failed to fetch project detail');
  const result = await response.json();
  return result.data || result;
};

// Create project
export const createProject = async (data: CreateProjectRequest): Promise<ProjectResponse> => {
  console.log('Creating project with data:', data);
  
  const response = await fetch(`${API_BASE_URL}/projects`, {
    method: 'POST',
    headers: getAuthHeaders(),
    credentials: 'include',
    body: JSON.stringify(data),
  });
  
  console.log('Create project response status:', response.status);
  
  if (!response.ok) {
    const errorText = await response.text();
    console.error('Create project failed:', {
      status: response.status,
      statusText: response.statusText,
      body: errorText,
    });
    
    try {
      const errorData = JSON.parse(errorText);
      throw new Error(errorData.message || errorData.error || 'Failed to create project');
    } catch (e) {
      throw new Error(`HTTP ${response.status}: ${errorText || response.statusText}`);
    }
  }
  
  const result = await response.json();
  console.log('Create project result:', result);
  // Backend returns ApiResponse wrapper: { success, message, data }
  return result.data || result;
};

// Update project
export const updateProject = async (data: UpdateProjectRequest): Promise<ProjectResponse> => {
  const response = await fetch(`${API_BASE_URL}/projects/${data.projectId}`, {
    method: 'PUT',
    headers: getAuthHeaders(),
    credentials: 'include',
    body: JSON.stringify(data),
  });
  if (!response.ok) throw new Error('Failed to update project');
  const result = await response.json();
  return result.data || result;
};

// Vote for project
export const voteProject = async (projectId: number): Promise<void> => {
  const token = localStorage.getItem('accessToken');
  const headers: HeadersInit = {};
  if (token) {
    headers['Authorization'] = `Bearer ${token}`;
  }
  
  const response = await fetch(`${API_BASE_URL}/projects/${projectId}/vote`, {
    method: 'POST',
    headers,
    credentials: 'include',
  });
  if (!response.ok) throw new Error('Failed to vote project');
};

// Unvote project
export const unvoteProject = async (projectId: number): Promise<void> => {
  const token = localStorage.getItem('accessToken');
  const headers: HeadersInit = {};
  if (token) {
    headers['Authorization'] = `Bearer ${token}`;
  }
  
  const response = await fetch(`${API_BASE_URL}/projects/${projectId}/unvote`, {
    method: 'DELETE',
    headers,
    credentials: 'include',
  });
  if (!response.ok) throw new Error('Failed to unvote project');
};

// Get voting projects
export const getVotingProjects = async (page: number = 0, size: number = 10): Promise<ProjectListResponse> => {
  const response = await fetch(`${API_BASE_URL}/projects/voting?page=${page}&size=${size}`, {
    method: 'GET',
    credentials: 'include',
  });
  if (!response.ok) throw new Error('Failed to fetch voting projects');
  const result = await response.json();
  const pageData = result.data || result;
  return {
    projects: pageData.content || [],
    totalPages: pageData.totalPages || 0,
    totalElements: pageData.totalElements || 0,
    currentPage: pageData.number || 0,
  };
};

File: be/ConnectEXE/src/main/java/com/connectexe/ConnectEXE/project/service/ProjectService.java
package com.connectexe.ConnectEXE.project.service;

import com.connectexe.ConnectEXE.entity.Project;
import com.connectexe.ConnectEXE.project.dto.request.CreateProjectRequest;
import com.connectexe.ConnectEXE.project.dto.request.UpdateProjectRequest;
import com.connectexe.ConnectEXE.project.dto.response.ProjectResponse;

import org.springframework.data.domain.Page;

import java.util.List;

public interface ProjectService {
    
    /**
     * Get random projects from subscribed users
     * @param userId Current user ID
     * @param limit Maximum number of projects to return
     * @return List of random projects
     */
    List<ProjectResponse> getRandomProjectsFromSubscriptions(String userId, int limit);
    
    /**
     * Get successful projects (status = SUCCESSFUL)
     * @param page Page number (0-indexed)
     * @param size Page size
     * @param userId Current user ID (optional, for checking vote status)
     * @return Page of successful projects
     */
    Page<ProjectResponse> getSuccessfulProjects(int page, int size, String userId);
    
    /**
     * Get all public projects with pagination
     * @param page Page number (0-indexed)
     * @param size Page size
     * @param userId Current user ID (optional, for checking vote status)
     * @return Page of projects
     */
    Page<ProjectResponse> getAllProjects(int page, int size, String userId);
    
    /**
     * Get projects by category
     * @param category Project category
     * @param page Page number (0-indexed)
     * @param size Page size
     * @param userId Current user ID (optional, for checking vote status)
     * @return Page of projects in the specified category
     */
    Page<ProjectResponse> getProjectsByCategory(Project.ProjectCategory category, int page, int size, String userId);
    
    /**
     * Get project detail by ID
     * @param projectId Project ID
     * @param userId Current user ID (optional, for checking vote status)
     * @return Project details
     */
    ProjectResponse getProjectDetail(String projectId, String userId);
    
    /**
     * Get projects sorted by vote count for voting page
     * @param page Page number (0-indexed)
     * @param size Page size
     * @param userId Current user ID (optional, for checking vote status)
     * @return Page of projects with vote counts
     */
    Page<ProjectResponse> getVotingProjects(int page, int size, String userId);
    
    /**
     * Create a new project
     * @param request Project creation request
     * @param userId ID of the user creating the project
     * @return Created project response
     */
    ProjectResponse createProject(CreateProjectRequest request, String userId);
    
    /**
     * Update an existing project
     * @param projectId Project ID
     * @param request Project update request
     * @param userId ID of the user updating the project
     * @return Updated project response
     */
    ProjectResponse updateProject(String projectId, UpdateProjectRequest request, String userId);
    
    /**
     * Vote for a project
     * @param projectId Project ID
     * @param userId ID of the user voting
     */
    void voteProject(String projectId, String userId);
    
    /**
     * Remove vote from a project
     * @param projectId Project ID
     * @param userId ID of the user removing vote
     */
    void unvoteProject(String projectId, String userId);
}
